File: artifacts/api-server/src/env/state.ts
/**
 * Environment state management for the OpenEnv RL environment.
 * Tracks the agent's position in the simulated website navigation.
 */

export type PageType = "home" | "search_results" | "product_detail" | "cart";

export interface HistoryEntry {
  page: PageType;
  action: string;
  current_product_id: string | null;
  step: number;
}

export interface EnvironmentState {
  session_id: string;
  task_id: string;
  current_page: PageType;
  search_results: string[];
  current_product_id: string | null;
  cart_items: string[];
  last_search_query: string | null;
  step_count: number;
  total_reward: number;
  is_done: boolean;
  history: HistoryEntry[];
  created_at: string;
  ended_at: string | null;
}

export function createInitialState(
  sessionId: string,
  taskId: string,
): EnvironmentState {
  return {
    session_id: sessionId,
    task_id: taskId,
    current_page: "home",
    search_results: [],
    current_product_id: null,
    cart_items: [],
    last_search_query: null,
    step_count: 0,
    total_reward: 0,
    is_done: false,
    history: [],
    created_at: new Date().toISOString(),
    ended_at: null,
  };
}

File: env/state.py
"""
Environment state definitions for the OpenEnv AI Web Navigation Training Environment.
"""

from dataclasses import dataclass, field
from typing import Optional, List
from datetime import datetime


PAGE_TYPES = ["home", "search_results", "product_detail", "cart"]


@dataclass
class HistoryEntry:
    page: str
    action: str
    current_product_id: Optional[str]
    step: int


@dataclass
class EnvironmentState:
    session_id: str
    task_id: str
    current_page: str = "home"
    search_results: List[str] = field(default_factory=list)
    current_product_id: Optional[str] = None
    cart_items: List[str] = field(default_factory=list)
    last_search_query: Optional[str] = None
    step_count: int = 0
    total_reward: float = 0.0
    is_done: bool = False
    history: List[HistoryEntry] = field(default_factory=list)
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())
    ended_at: Optional[str] = None


def create_initial_state(session_id: str, task_id: str) -> EnvironmentState:
    """Create a fresh environment state for a new session."""
    return EnvironmentState(session_id=session_id, task_id=task_id)
